### ISSUE TITLE
[Android] View order (z-index) is not respected for <Surface>, Native Views always get covered up

### ISSUE THREAD

**Issue by gr4yscale** [ProjectSeptemberInc/gl-react-native#57]
When attempting to lay native views on top of a surface using JSX similar to the below, we see that on iOS native views render on top of the surface. On Android, no matter what styling you use (I've tried position: 'absolute' on Surfaces, native views, etc), the surface renders above everything else.

```
  <View style={styles.container}>
    <Surface
        width={deviceWidth}
        height={300}
        opaque={false}
        ref="helloGL"
        pixelRatio={2}
    >
      <HelloGL />
    </Surface>
    <Image style={{position:'absolute', top: 100, left: 30, width: 300, height: 100}} source={{ uri: 'http://i.imgur.com/S22HNaU.png'}} width={300} height={100} />
    <Text style={{position:'absolute', top: 100, left: 30, width: 300, height: 100,backgroundColor:'transparent', color:'#FFFFFF', fontSize:36}}>SOME NATIVE TEXT ABOVE AN IMAGE, ABOVE A SHADER</Text>
  </View>
```

iOS screenshot:

![img_8381](https://cloud.githubusercontent.com/assets/1342396/12446883/9bb4987a-bf64-11e5-8163-92e5d31c31f7.PNG)

Android screenshot:

![screenshot_2016-01-20-10-51-39](https://cloud.githubusercontent.com/assets/1342396/12446889/a6b63012-bf64-11e5-9416-e56fd30e5409.png)

On Android with the same JSX as above, except replacing the the <Surface> with a native view styled with a red backgroundColor:

![img_8383](https://cloud.githubusercontent.com/assets/1342396/12448785/50b0f330-bf70-11e5-8c51-8658fdc39eb4.JPG)

**Comment by gr4yscale** [comment]
@gre when you said "It might be related to the <View><GLCanvas> specific internal nesting" did you mean in gl-react, or in react-native?

**Comment by gre** [comment]
> @gre when you said "It might be related to the specific internal nesting" did you mean in gl-react, or in react-native?

somehow between the 2 actually :smile: 

**gl-react** provides [`createSurface`](https://github.com/ProjectSeptemberInc/gl-react/blob/master/src/createSurface.js#L23) that implementations like **gl-react-native** can use to create [`Surface`](https://github.com/ProjectSeptemberInc/gl-react-native/blob/master/src/Surface.js#L13-L47) with 3 rendering functions: `renderVcontainer` which takes in parameter the result of `renderVGL` and multiple calls to `renderVcontent` and create this nesting:

``` html
<View ...>
  { _contents_.map( content =>
    <View ...>{ content }</View>
  }
  <GLCanvas ...>
</View>
```

`_contents_` here in an array of all **virtual** views you have passed in uniforms that will be rasterized by `GLCanvas`. I had to put it here so it's actually rendered as ObjC **native** views and inside the context where you will use that `Surface`. GLCanvas impl will do a **glCanvas.parentView.children[ index ]** to access a content to rasterize.

> I've named it `_contents_` because it's not **exactly** the same impl, but was just to explain it easily.

**Comment by gre** [comment]
just dropping a thought on the raised issue: I'm wondering if it can be related to http://facebook.github.io/react-native/docs/view.html#collapsable , should see if forcing `collapsable={false}` on the root View (of that Surface `renderVcontainer` impl) would help

**Comment by gr4yscale** [comment]
I tried the collapsable={false} change and didn't see any better results =\

```
function renderVcontainer ({ style, width, height, visibleContent, eventsThrough }, contents, renderer) {
  const parentStyle = {
    position: "relative",
    ...style,
    width: width,
    height: height,
    overflow: "hidden",
  };
  return <View
    pointerEvents={!visibleContent && eventsThrough ? "none" : "auto"}
    collapsable={false}
    style={parentStyle}>
    {contents}
    {renderer}
  </View>;
}
```

Started digging into the android GLCanvas native module trying to make sense of it - is it _always_ rasterizing a bitmap snapshot (https://github.com/ProjectSeptemberInc/gl-react-native/blob/d5def9d7b7c8c56444e35778ab8bd9e941f0ce0c/android/src/main/java/com/projectseptember/RNGL/GLCanvas.java#L329) of native views and passing them into the framebuffer as textures?

Sorry if these are silly questions, I'm relatively inexperienced with OpenGL (and Android) but keen to learn and am super excited for React Native + GL!

**Comment by gr4yscale** [comment]
Found it! 

https://github.com/ProjectSeptemberInc/gl-react-native/blob/d5def9d7b7c8c56444e35778ab8bd9e941f0ce0c/android/src/main/java/com/projectseptember/RNGL/GLCanvas.java#L92

Wooo!

**Comment by gr4yscale** [comment]
I opened a PR, but you'll probably want to test it thoroughly before merging.

With the change, my native views render over the GLCanvas / GLSurfaceView, but I've got an issue now when trying to navigate to another screen (updating a react components child components, I believe) I end up with half of the screen black:

![screenshot_2016-01-22-11-22-40](https://cloud.githubusercontent.com/assets/1342396/12509511/e48b69b0-c0fa-11e5-8d31-679b1044e332.png)

![screenshot_2016-01-22-11-22-43](https://cloud.githubusercontent.com/assets/1342396/12509516/e98900c6-c0fa-11e5-8e8b-cdd18f943938.png)

Pardon the silly colors, hah - just prototyping!

**Comment by gre** [comment]
Hey @gr4yscale , looks like you are having fun with this background :)

the black area is weird, what's in this area?

also just a perf feedback, make sure you put the Surface inside a StaticContainer  (`require("StaticContainer.react")`) or in your own component where you implement `shouldComponentUpdate`, that way you are sure it doesn't renders everytime (that said, maybe React Native is already doing his job. I don't think the setters get called if nothing changes, but I'm not sure on Android side)

**Comment by gr4yscale** [comment]
I realized what the black area is - that's where the keyboard was previously. After it's dismissed before swapping out child components I guess there is no redraw - it's fine after the keyboard comes back up and then goes away again... I'm not sure how to force a redraw on the RN side.

Thanks for the tip StaticContainer, I had just set shouldUpdateComponent() to return false in the component that rendered the <Surface>

**Comment by gre** [comment]
alright, have fun :)

issue fixed by #59
